### PR TITLE
CNI networking installation support

### PR DESCRIPTION
**Pull request by chrislovecnm**
Work on getting cni networks such a weave functioning.  I did encounter this probably https://github.com/kubernetes/kubernetes/issues/20379

```
Oct  9 04:09:44 ip-172-20-111-165 kubelet[999]: E1009 04:09:44.019978     999 kubelet.go:2095] Failed killing the pod "cassandra-5": failed to "TeardownNetwork" for "cassandra-5_default" with TeardownNetworkError: "Failed to teardown network for pod \"b4f4f534-8dd5-11e6-a44d-02c1f7bba931\" using network plugins \"cni\": no such file or directory"
```

This is now tested an ready for master.  I will be testing an install above the 50 node limit to fully vet.

### CODE REVIEW

**Comment by chrislovecnm** [comment]
Yah noticed that too, just modeled the original code. @justinsb?

File: upup/models/config/components/kubelet/_networking_cni/_k8s_1_4/kubelet.cni.1_4.options [hunk 1]
```diff
@@ -0,0 +1,8 @@
+Kubelet:
+  # AWS MTU is 9001
+  NetworkPluginMTU: 9001
+
+# TODO: Having to duplicate MasterKubelet & Kubelet feels wrong
```
Comment (kris-nova) [inline upup/models/config/components/kubelet/_networking_cni/_k8s_1_4/kubelet.cni.1_4.options]: What feels wrong here? Generally vagrant `TODO`'s make me at least want to raise a question :)

File: upup/pkg/fi/cloudup/tagbuilder.go [hunk 2]
```diff
@@ -23,6 +23,9 @@ func buildCloudupTags(cluster *api.Cluster) (map[string]struct{}, error) {
 		// external is based on kubenet
 		tags["_networking_kubenet"] = struct{}{}
 		tags["_networking_external"] = struct{}{}
+	} else if networking.CNI != nil {
+		// external is based on cni, weave, flannel, etc
+		tags["_networking_cni"] = struct{}{}
```
Comment (kris-nova) [inline upup/pkg/fi/cloudup/tagbuilder.go]: More empty structs? Are these just place holders for future members or something?

File: upup/models/config/components/kubelet/_networking_cni/_k8s_1_4/kubelet.cni.1_4.options (hunk 1 above)
Comment (chrislovecnm) [inline upup/models/config/components/kubelet/_networking_cni/_k8s_1_4/kubelet.cni.1_4.options]: Comments ported over from the files I stole from kubenet. I can remove.

File: upup/pkg/fi/cloudup/tagbuilder.go (hunk 2 above)
Comment (chrislovecnm) [inline upup/pkg/fi/cloudup/tagbuilder.go]: Also answered in French. Again just modeling code done before. We can cleanup if you care to recommend a pattern.

**Comment by chrislovecnm** [comment]
should solve https://github.com/kubernetes/kops/issues/537 as well

**Comment by chrislovecnm** [comment]
Should solve https://github.com/kubernetes/kops/issues/315 as well

**Comment by WillPlatnick** [comment]
I am so excited for this. Thanks @chrislovecnm

**Comment by chrislovecnm** [comment]
@WillPlatnick looking good, just needing more testing.

File: upup/models/config/components/kubelet/_networking_cni/_k8s_1_5/kubelet.cni.1_5.options [hunk 3]
```diff
@@ -0,0 +1,7 @@
+Kubelet:
+  # AWS MTU is 9001
+  NetworkPluginMTU: 9001
```
Comment (kzwang) [inline upup/models/config/components/kubelet/_networking_cni/_k8s_1_5/kubelet.cni.1_5.options]: Hi @chrislovecnm , I got a question about this MTU. According to Weave document, the MTU should be 8950 as there is an overhead of 50 bytes for Vxlan, will this be handled automatically? 

Another question is I can see `kubenet` is using the MTU passed to the plugin (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/network/kubenet/kubenet_linux.go#L128), however, for cni it doesn't seem it's using the MTU value (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/network/cni/cni.go#L144). Is this been handled somewhere else?

Comment (chrislovecnm) [inline upup/models/config/components/kubelet/_networking_cni/_k8s_1_5/kubelet.cni.1_5.options]: I am going to remove it.  Thanks

**Comment by offlinehacker** [comment]
We are testing on new cluster, and it seems to work without problems so far, i also did some basic policy testing which also seems to work.

**Comment by chrislovecnm** [comment]
@offlinehacker I am getting odd log message when I delete pods.  https://github.com/weaveworks/weave-kube/issues/32

**Comment by chrislovecnm** [comment]
I need more testing and squash my commits prettier :)

**Comment by chrislovecnm** [comment]
@JuJu227 if you could take a look, specifically at the documentation, would be much appreciated.

File: upup/pkg/fi/cloudup/populatecluster_test.go [hunk 4]
```diff
@@ -145,6 +145,45 @@ func TestPopulateCluster_Kubenet(t *testing.T) {
...
+		NonMasqueradeCIDR: c.Spec.NonMasqueradeCIDR,
+		CloudProvider:     c.Spec.CloudProvider,
+	}
+
+	/*
+		c.Spec.Zones = []*api.ClusterZoneSpec{
+			{Name: "us-west-2a", CIDR: "172.20.2.0/27"},
+			{Name: "us-west-2b", CIDR: "172.20.2.32/27"},
+			{Name: "us-west-2c", CIDR: "172.20.2.64/27"},
+		}*/
```
Comment (JuJu227) [inline upup/pkg/fi/cloudup/populatecluster_test.go]: Suggest removing dead code.

**Comment by kris-nova** [comment]
@chrislovecnm - this looks good.. From a code/documentation standpoint this all makes sense. Will have to defer to you and @justinsb for functionality success.

File: upup/pkg/fi/cloudup/populatecluster_test.go (hunk 4 above)
Comment (chrislovecnm) [inline upup/pkg/fi/cloudup/populatecluster_test.go]: Will do

File: upup/pkg/fi/cloudup/tagbuilder.go (hunk 2 above)
Comment (justinsb) [inline upup/pkg/fi/cloudup/tagbuilder.go]: This is the "best way" to do sets in golang

**Comment by justinsb** [comment]
LGTM... found a few problems:

* We hit https://github.com/kubernetes/kubernetes/issues/31369 when using a machine with a single core - m4.large fixes it.

* Have to SSH to the master to install the weave DS

* When a node first comes up it isn't really ready; this self heals though:

```
  1m    1m      1       {kubelet ip-172-20-68-75.ec2.internal}          Warning FailedSync      Error syncing pod, skipping: failed to "SetupNetwork" for "guestbook-03xsn_default" with SetupNetworkError: "Failed to setup network for pod \"guestbook-03xsn_default(4c4a9f7b-9676-11e6-b502-061a7a80de7a)\" using network plugins \"cni\": unable to allocate IP address: Post http://127.0.0.1:6784/ip/416e5685e89795642fa79a0959dfbf687aa452fdba527c0ff11ad8dd1519c652: dial tcp 127.0.0.1:6784: getsockopt: connection refused; Skipping pod"
```

* If we can change the pod IP range away from 10.0.0.0/8 that will avoid conflicts with EC2 classic.  We default to 100.64.x.x, if we can honor that it's probably best

*  At 125 nodes, started seeing this error:
```
E1020 03:58:52.832357       5 endpoints_controller.go:514] Endpoints "guestbook" is invalid: [subsets[0].addresses[52].nodeName: Forbidden: Cannot change NodeName for 10.38.0.1 to ip-172-20-92-169.ec2.internal, subsets[0].addresses[97].nodeName: Forbidden: Cannot change NodeName for 10.42.128.1 to ip-172-20-91-115.ec2.internal]
```

**Comment by justinsb** [comment]
Merging - we can address those issues post merge!

**Comment by justinsb** [comment]
@chrislovecnm do we have issues to track the the 5 snags above?

**Pull request merged**